### ISSUE TITLE
python urllib3 build cripples metadata

### ISSUE THREAD

**Issue by stesie** [NixOS/nixpkgs#46318]
## Issue description

I noticed, while packaging azure-cli, that Nix' build process "somehow" cripples the dependency metadata of `urllib3` package... that is it removes python_version ranges from "Requires-Dist" entries (causing build of dependencies to fail).

Original metadata entries:

```
Requires-Dist: pyOpenSSL>=0.14,<18.0.0; python_version=="2.6" and extra == 'secure'
Requires-Dist: pyOpenSSL>=0.14; python_version=="2.7" and extra == 'secure'
Requires-Dist: cryptography>=1.3.4; python_version<="2.7" and extra == 'secure'
Requires-Dist: idna>=2.0.0; python_version<="2.7" and extra == 'secure'
Requires-Dist: certifi; extra == 'secure'
Requires-Dist: ipaddress; python_version<="2.7" and extra == 'secure'
Requires-Dist: PySocks>=1.5.6,<2.0,!=1.5.7; extra == 'socks'
```

$ nix-shell -p python36Packages.urllib3
$ grep -e Requires-Dist $buildInputs/lib/python3.6/site-packages/urllib3-1.23.dist-info/METADATA
```
Requires-Dist: pyOpenSSL (>=0.14); extra == 'secure'
Requires-Dist: cryptography (>=1.3.4); extra == 'secure'
Requires-Dist: idna (>=2.0.0); extra == 'secure'
Requires-Dist: certifi; extra == 'secure'
Requires-Dist: ipaddress; extra == 'secure'
Requires-Dist: PySocks (!=1.5.7,<2.0,>=1.5.6); extra == 'socks'
```

This causes problems if another package requires `urllib3[secure]`, say on python 3.6.  Normally that package would not need e.g. `ipaddress`, since it's only needed for python <= 2.7 ... Yet with the check being removed it suddenly depends on `ipaddress` which isn't available on python 3.6  (and doesn't make sense since ipaddress is part of core now)

### Steps to reproduce

Download current urllib3 source package from https://files.pythonhosted.org/packages/3c/d2/dc5471622bd200db1cd9319e02e71bc655e9ea27b8e0ce65fc69de0dac15/urllib3-1.23.tar.gz and see `setup.cfg` in root.

Then
$ nix-shell -p python36Packages.urllib3
$ grep -e Requires-Dist $buildInputs/lib/python3.6/site-packages/urllib3-1.23.dist-info/METADATA

... and see how it is broken

Also, if you'd like to try with a dependent package, try azure-cli branch from https://github.com/stesie/nixpkgs/commits/azure-cli and build `nix-build  -A python36Packages.azure-cli-appservice`.  You need to revert the "urllib3: install from wheel package format" workaround commit before, for it to fail:

```
Collecting ipaddress; extra == "secure" (from urllib3[secure]>=1.18->azure-cli-appservice==0.1.36)
  Could not find a version that satisfies the requirement ipaddress; extra == "secure" (from urllib3[secure]>=1.18->azure-cli-appservice==0.1.36) (from versions: )
No matching distribution found for ipaddress; extra == "secure" (from urllib3[secure]>=1.18->azure-cli-appservice==0.1.36)
```

## Potential fix/workaround

Instead of installing urllib3 from sdist it can be installed from wheel, with patch https://github.com/stesie/nixpkgs/commit/b0db101200f53c8aac22aea6a8d514e9ce3e1559

But I'm unsure if this is a proper fix ... or if there's other packages that are possibly affected by the same problem

## Technical details

 - system: `"x86_64-linux"`
 - host os: `Linux 4.14.68, NixOS, 19.03pre151837.ca2ba44cab4 (Koi)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.1`
 - channels(root): `"nixos-19.03pre151837.ca2ba44cab4"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`

/cc @FRidh

**Comment by FRidh** [comment]
We're not doing anything special in our build process regarding this, so if there's a bug, then it's most likely in `setuptools` or `wheel`.

**Comment by colemickens** [comment]
related? https://github.com/inveniosoftware/troubleshooting/issues/1

**Comment by colemickens** [comment]
I don't know much about python. Is it possible that one of the install methods uses setup.cfg and the other uses setup.py? That would explain why one works and not the other. IT *seems* like setup.cfg has the right data, but setup.py is missing the extra metadata that ipaddress is only needed on 2.x?

**Comment by colemickens** [comment]
Interesting. I learned more than I wanted to about python packaging. I've made a small local change following hints in the thread I linked, and I've got urllib3 generating what appears to be correct `METADATA` now. It looks like an upstream issue with `urllib3`.

I'll do some additional looking tomorrow and file an Issue or send a PR for urllib3.

In the meantime, should we install via the other method to workaround for nixpkgs?

**Comment by colemickens** [comment]
Filed upsteam issue: https://github.com/urllib3/urllib3/issues/1460

**Comment by colemickens** [comment]
Upstream issue was resolved. @stesie want to give it another shot?

**Comment by stesie** [comment]
@colemickens noticed it yesterday, ... waiting for them to publish a release.  I think it's not worth packaging current git master for nixpkgs.

**Comment by colemickens** [comment]
@stesie Oh, my mistake, I'd assumed it was closed as part of a release. I see you're on top of it! Thanks!

**Comment by dotlambda** [comment]
We can try applying https://github.com/urllib3/urllib3/commit/f80ff34bffb2018528ccd2ccf82e533d321f2af8 as a patch if you want.

**Comment by flokli** [comment]
We'd need to bump urllib3 to 1.25.x, which includes the fix.

**Comment by flokli** [comment]
I assume this is fixed, as `urllib3` is on 1.25.9 on master. Closing, feel free to reopen/comment if I'm wrong.
